include/deal.II/base/aligned_vector.h

```diff
@@ -1036,6 +1124,405 @@ AlignedVector<T>::fill(const T &value)
+                                   shmem_group_communicator,
+                                   &base_ptr,
+                                   &shmem_window);
+    AssertThrowMPI(ierr);
+
+    ierr = MPI_Info_free(&mpi_info);
+    AssertThrowMPI(ierr);
+  }
+
+
+  // **** Step 4 ****
+  // The next step is to teach all non-shmem root processes what the pointer to
+  // the array is that the shmem-root created. MPI has a nifty way for this
+  // given that only a single process actually allocated memory in the window:
+  // When calling MPI_Win_shared_query, the MPI documentation says that
+  // "When rank is MPI_PROC_NULL, the pointer, disp_unit, and size returned are
+  // the pointer, disp_unit, and size of the memory segment belonging the lowest
+  // rank that specified size > 0. If all processes in the group attached to the
+  // window specified size = 0, then the call returns size = 0 and a baseptr as
+  // if MPI_ALLOC_MEM was called with size = 0."
+  //
+  // This will allow us to obtain the pointer to the shmem root's memory area,
+  // which is the only we one we care about. (None of the other processes have
+  // even allocated any memory.) But this will also retrieve the shmem root's
+  // disp_unit, which in step 3 above we have abused to pass along the number of
+  // elements in the array.
+  //
+  // We don't need to do this on the shmem root process: This process has
+  // already gotten its base_ptr correctly set above, and we can determine the
+  // array size by just calling size().
+  unsigned int array_size =
+    (is_shmem_root ? size() : numbers::invalid_unsigned_int);
+  if (is_shmem_root == false)
+    {
+      int       disp_unit;
+      MPI_Aint  alloc_size; // not actually used
+      const int ierr = MPI_Win_shared_query(
+        shmem_window, MPI_PROC_NULL, &alloc_size, &disp_unit, &base_ptr);
+      AssertThrowMPI(ierr);
+
+      // Make sure we actually got a pointer, and also unpack the array size as
+      // discussed above.
+      Assert(base_ptr != nullptr, ExcInternalError());
+
+      array_size = disp_unit;
+    }
+
+
+  // **** Step 5 ****
+  // Now that all processes know the address of the space that is visible to
+  // everyone, we need to figure out whether it is properly aligned and if not,
+  // find the next aligned address.
+  //
+  // std::align does that, but it also modifies its last two arguments. The
+  // documentation of that function at
+  // https://en.cppreference.com/w/cpp/memory/align is not entirely clear, but I
+  // *think* that the following should do given that we do not use base_ptr and
+  // available_space any further after the call to std::align.
+  std::size_t available_space       = alloc_size;
+  void *      base_ptr_backup       = base_ptr;
+  T *         aligned_shmem_pointer = static_cast<T *>(
+    std::align(align_by, array_size * sizeof(T), base_ptr, available_space));
+  Assert(aligned_shmem_pointer != nullptr, ExcInternalError());
```

Comment:
Hmmm... I would think that the allocation is actually shifted so using `base_ptr_backup` would potentially point to something wrong if `std::align` actually does something.

Reply:
Can you elaborate? I *think* what you're saying is that `std::align` actually shifts data around (from `base_ptr` to the aligned address), but I don't think that is what `std::align` does. My reading is that it really just aligns the *pointer*, but doesn't touch data. Regardless, we haven't actually put any data into the memory region, so that is not a concern.

Or am I completely misreading what you're saying?

Reply:
Yes, I agree that it only changes the value of the pointer and you're right that we don't have any data we should worry about at this point.